New commands to output results to file and database and New Power Bi File

SQLDBAWithABeard opened the pull request:
OK This is a big change, so I would like some feedback on this before it gets merged.

There are some bug fixes to checks that I found along the way but the important stuff is

## New Commands

I have added 

- Convert-DbcResult
    This will take the out put of `Invoke-DbcCheck` WITH THE `-PassThru` parameter :-) and convert the output to a PowerShell object that looks like this

````
Date           : 27/12/2019 13:40:20
Label          : Morning-Checks
Describe       : Future File Growth
Context        : Testing for files likely to grow soon
Name           : Database msdb file MSDBData has free space under threshold
Database       : msdb
ComputerName   : SQL2019N5
Instance       : SQL2019N5\DAVE
Result         : Failed
FailureMessage : Expected the actual value to be greater than or equal to 10, because free space within the file should be lower than threshold of 10 %, but got 5.932.
````
The label is a way of defining a group of checks (like the environmenttag for Update-DbcPowerBiDataSource)

You can try it with

````
Invoke-DbcCheck -SqlInstance INSTANCE -Check CHECK -PassThru | Convert-DbcResult -Label Some-Label
````
- Set-DbcFile
     This command will take an object from `Convert-DbcResult` and either save to a file or add to a file of type CSV, JSON or XML

You can try it with

````
Invoke-DbcCheck -SqlInstance SQLINSTANCE -Check AutoClose -Show Fails -PassThru|Convert-DbcResult -Label 'Some-Label' | Set-DbcFile -FilePath C:\temp\dbachecks\ -FileName Auto-close.json -FileType Json 
````

or by setting the  output of the checks to a variable and piping that

````
$Tests = Invoke-DbcCheck -Check AutoClose -Show Fails -PassThru
$Tests |Convert-DbcResult -Label 'Some-Label' | Set-DbcFile -FilePath C:\temp\dbachecks\ -FileName Auto-close.json -FileType Json 
$Tests |Convert-DbcResult -Label 'Some-Label' | Set-DbcFile -FilePath C:\temp\dbachecks\ -FileName Auto-close.csv -FileType CSV 
$Tests |Convert-DbcResult -Label 'Some-Label' | Set-DbcFile -FilePath C:\temp\dbachecks\ -FileName Auto-close.xml -FileType XML
$Tests |Convert-DbcResult -Label 'Some-Label' | Set-DbcFile -FilePath C:\temp\dbachecks\ -FileName Auto-close.json -FileType Json -Force
$Tests |Convert-DbcResult -Label 'Some-Label' | Set-DbcFile -FilePath C:\temp\dbachecks\ -FileName Auto-close.csv -FileType CSV -Force 
$Tests |Convert-DbcResult -Label 'Some-Label' | Set-DbcFile -FilePath C:\temp\dbachecks\ -FileName Auto-close.xml -FileType XML -Force
$Tests |Convert-DbcResult -Label 'Some-Label' | Set-DbcFile -FilePath C:\temp\dbachecks\ -FileName Auto-close.json -FileType Json -Append
$Tests |Convert-DbcResult -Label 'Some-Label' | Set-DbcFile -FilePath C:\temp\dbachecks\ -FileName Auto-close.csv -FileType CSV -Append
$Tests |Convert-DbcResult -Label 'Some-Label' | Set-DbcFile -FilePath C:\temp\dbachecks\ -FileName Auto-close.xml -FileType XML -Append
````

- Write-DbcTable
    This command will take the output of `Convert-DbcResult` and write it to a database table - It will also take the output of `Get-DbcCheck` and place that in a table called dbachecksChecks (This is required for the new pbix

You can try it with

````
$Tests = Invoke-DbcCheck -Check AutoClose -Show Fails -PassThru |Convert-DbcResult -Label 'Some-Label' | Write-DbcTable -SqlInstance SQL2017N5 -Database tempdb -Table SomeLabelChecks 
````

I have also added a new PowerBi to the repo under \bin which I would really like some clever people to make more beautiful!

Here is what I have so far - The data comes from the database created by `Write-DbcTable`

https://app.powerbi.com/reportEmbed?reportId=48d7902f-f3d6-4d46-9308-bce4aeda14eb&autoAuth=true&ctid=b122247e-1ebf-4b52-b309-c2aa7436fc6b&config=eyJjbHVzdGVyVXJsIjoiaHR0cHM6Ly93YWJpLW5vcnRoLWV1cm9wZS1yZWRpcmVjdC5hbmFseXNpcy53aW5kb3dzLm5ldC8ifQ%3D%3D

The first page shows the information for the latest date in the data

![image](https://user-images.githubusercontent.com/6729780/71584555-f3501300-2b0a-11ea-8ed1-1535e0c4d7f4.png)


The second page

![image](https://user-images.githubusercontent.com/6729780/71584571-06fb7980-2b0b-11ea-90b6-4c880ff6ca6e.png)


shows the info by date

The third page by database

![image](https://user-images.githubusercontent.com/6729780/71584581-18448600-2b0b-11ea-8d73-0e15c8e135dc.png)


The fourth page is the drill through page showing the information for the checks enabling you to see the failure messages

![image](https://user-images.githubusercontent.com/6729780/71584653-85581b80-2b0b-11ea-95c5-41ca22ac43a9.png)

![image](https://user-images.githubusercontent.com/6729780/71584666-9739be80-2b0b-11ea-9d4a-a515a2deb79f.png)

![image](https://user-images.githubusercontent.com/6729780/71584701-cf410180-2b0b-11ea-84a2-ee0f573aef89.png)

![image](https://user-images.githubusercontent.com/6729780/71584720-e5e75880-2b0b-11ea-8d27-911f9a5fe052.png)

This is similar to another report that I build so that DBAs could quickly get information about the checks and the reasons for failures and identify what to work on, DBA team leaders could identify risk areas and see how the team were doing and Management could see pretty pictures which changed when they clicked on them.

I am interested in making these reports as gorgeous as possible with the info that people expect to see, so please feel free to comment or to take the Power Bi file new-dbachecks.pbix in the bin folder and make it more beautiful

~~Some items are not finished yet, the Pester Tests wont succeed as there is no help for the new commands either but if you wish to try this out.~~

Tests now pass.
If you want to try this out

Clone the repo or update your own clone with 
````
Git status
git fetch upstream
Git checkout development
git merge upstream/development
git checkout New-Json
````
Then navigate to the root of your repo and run 

````
ipmo dbachecks.psd1
````
to load the new code

wsmelton commented:
..mind you I have not heavily used or looked at code, so just a thought.

Would `ConvertTo-DbaDataTable` not be of use to you here to simplify code?

I would see the commands be like `ConvertTo-DbcJson` or `ConvertTo-DbcCsv` as well to just output that data in the format expected. It prevents the need for having the `Set-DbcFile` command and allows them to just use `Out-File` or send that data to anything else they want. Breaking it up into separate convert commands would also make them easier to maintain and test as well...from my thought process.

SQLDBAWithABeard commented:
@wsmelton -we could use ConvertToDbcDataTable but we need Convert-DbcResult to do more than just convert the output to a datatable (although it does!), we parse the results for 
Instance, Database and enable adding a label so different configurations can be easily identified

 I did think about using Set-DbcFile or writing three commands for each type of file and settled on one command - I am not precious about it, it could easily be changed if you think it wise

SQLDBAWithABeard commented:
Tests now pass :-)

marcingminski commented:
Hi, I was looking at the layout changes we twitted about and whilst doing it, I noticed the following:

- Day in the Dates dimension should be of date format, not text. This is to allow PBI to use data based filtering (relative dates) and to use continous axis rather than category.
- Dates dimension should be limited to the dates in the actual data set - the dimension itself currently starts at 1899 and as is dynamically calculated (measure) it takes a bit of resource to actually get to todays date. Or, if there is a real need to show dates without data i.e. to show days where evaluation did not happen, it should be a static table joined in the power query before it gets to DAX.
- There should probably be a server dimension but I don't see the schema - I need to deploy dbachecks in my env for this to validate.
- From performnace point of view, there should be Check dimension and Check results. It may not matter if the result sets are small (i.e. less than few thousand rows) but I'd be keen to implement it for completness.

I can change and fix all of that but it may take a little while so not sure how long you want to wait with the PR. There may also be more tweaks to do as I go along, I only had a brief look.

As far as the layout changes go, this is obviously purely personal preference and just a suggetion. I looked at it from my experience as a dba, where I managed over 3k instances. What I wanted to know on daily basis:
- is the enviroment getting worse or better
- what are the key issues 
- what are the servers with most failures
- what has changed since last evaluation or since the baseline (this is tricky to do with the current set so not yet on the dashboard)

Then, look at the standard daily issues (backups, agent failures), then on weekly and monthy basis look at the non-standard issues, dbs with auto-close, shrink anything that may need a change request to implement etc.

I also noticed there is no green color in the existing theme, which is probably what you said about it being color-blind friendly. I therefore used colours present in the current theme - blue instead of green and the existing reddish-orangeish. I assume this is color-blind friendly color? If not, happy to change. 

With this in mind, I have rolled most of the data into one tab, with drill throughs and drill downs
![image](https://user-images.githubusercontent.com/8288333/71549406-b245ea80-29b4-11ea-8e60-0e59c045c002.png)

SQLDBAWithABeard commented:
Thank you @marcingminski :-)

The Dates dimension is created with calendarauto which I thought would create the dates for the dates in the model - I don't know why it goes back to 1899 :-( - Need a specialist to understand that one!!

I have updated the report with some of your suggestions :-) and updated the screenshots above

I want the front page to be the latest checks for quick access and the second page to be the trending as I want this first for the DBA who is coming in in the morning knowing what to fix. 

Of course, anyone can set the first page that is seen as they want ;-)

The dimensions for Instance, database etc are a stage 2 improvement :-) - I plan on creating using the table that is being used for the reports as a staging table and then having a sproc to update the "proper" table

SQLDBAWithABeard commented:
Hey @jpomfret - Can you check the logic is correct for [4bf1374](https://github.com/sqlcollaborative/dbachecks/pull/727/commits/4bf1374ddc77b870fbb6864ae90d15d4a97db0e6)

jpomfret commented:
Looks good to me @SQLDBAWithABeard, comparing the difference in this job run (as long as its in the last $maxDays) with the average run from the last $maxDays.
Tested out the query against a test instance and results looked good.

SQLDBAWithABeard commented:
Cool beans, Thank you @jpomfret - Got myself in a twizzle and wasnt sure if it was right in the end!!

marcingminski commented:
The CALENDARAUTO calculates ranges automatically, based on the data in the model, I guess it is not able to establigh what the min and max dates are in the data model, it goes all the way back to 1899 and in the future.

You could just use CALENDAR function and tell it your min and max dates from the check table:

```
Dates = 
VAR BaseCalendar =
    CALENDAR(  MIN( CheckResults[CheckDate] ), MAX(CheckResults[CheckDate]))
RETURN
    GENERATE (
        BaseCalendar,
        VAR BaseDate = [DATE]
        VAR YearDate = YEAR ( BaseDate )
        VAR MonthNumber = MONTH ( BaseDate )
        RETURN ROW (

            "Day", FORMAT(BaseDate, "mm/dd/yyyy"),
            "Year", YearDate,
            "Month Number", MonthNumber,
            "Month", FORMAT ( BaseDate, "MMM" ),
            "Month-Year", FORMAT ( BaseDate, "MMM-YY" )
        )
    )
```

![image](https://user-images.githubusercontent.com/8288333/71598555-30ce9380-2b3f-11ea-85e1-4ad7119509b0.png)

PsPsam commented:
I haven't tried this yet. but I would soo like to have this merged in to the master by just looking at it. Setting up dbachecks at a customer so that they can follow there compliance and this PR looks like a very good fit for them to follow there progress in getting more compliant to there standard.

SQLDBAWithABeard commented:
I have to see if I can merge all of the changes and bring this in. I might try at the weekend but no promises!

ClaudioESSilva commented on functions/Write-DbcTable.ps1:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    $SqlInstanceSmo = Connect-DbaInstance -SqlInstance $SqlInstance -SqlCredential $SqlCredential
+
```
Please see comment on line 73

I would add here something like this example:

``` powershell
# If specified table does not exists, create with specific datatypes to avoid nvarchar(max)
if ($null -eq (Get-DbaDbTable -Sqlinstance $SqlInstanceSmo -Database $database -Table "[$schema].[$table]")) {
    $sqlTableCreation = @"
CREATE TABLE [$schema].[$table](
    [Date] [datetime2](3) NOT NULL,
    [Label] [nvarchar](255) NULL,
    [Describe] [nvarchar](255) NULL,
    [Context] [nvarchar](255) NULL,
    [Name] [nvarchar](255) NULL,
    [Database] [nvarchar](255) NULL,
    [ComputerName] [nvarchar](255) NULL,
    [Instance] [nvarchar](255) NULL,
    [Result] [nvarchar](5) NULL,
    [FailureMessage] [nvarchar](255) NULL
) ON [PRIMARY] 
GO

CREATE CLUSTERED INDEX CI_DATE ON [$schema].[$table]
(
    [Date]
) WITH (DATA_COMPRESSION = PAGE) ON [PRIMARY]
GO
"@
    Invoke-DbaQuery -SqlInstance $SqlInstanceSmo -Database $Database -Query $sqlTableCreation
}
```

SQLDBAWithABeard replied:
Added this (with a couple of minor changes) @ClaudioESSilva

ClaudioESSilva commented on functions/Write-DbcTable.ps1:
```diff
@@ -0,0 +1,75 @@
+    $SqlInstanceSmo = Connect-DbaInstance -SqlInstance $SqlInstance -SqlCredential $SqlCredential
+
+    if ($PSCmdlet.ShouldProcess("$schema.$database" , "On $SqlInstance - Add or update the Configs tables in ")) {
+       Get-DbcCheck | Write-DbaDataTable -SqlInstance $SqlInstanceSmo -Database $Database -Table dbachecksChecks -Schema $Schema -AutoCreateTable -Truncate
```
Is the `$truncate` missing at the end of this line? `-Truncate:$truncate`
Will this table being truncated on every single run? Why?

SQLDBAWithABeard replied:
Could you take a look at the latest version - I havent put your changes in but I hadf 169 commits at home!!

SQLDBAWithABeard replied:
We will always truncate and reload the table holding the checks information as it only has a couple hundred rows

ClaudioESSilva commented on functions/Write-DbcTable.ps1:
```diff
@@ -0,0 +1,75 @@
+       Get-DbcCheck | Write-DbaDataTable -SqlInstance $SqlInstanceSmo -Database $Database -Table dbachecksChecks -Schema $Schema -AutoCreateTable -Truncate
+    }
+    if ($PSCmdlet.ShouldProcess("$Schema.$database" , "On $SqlInstance - Add dbachecks results to $Table in")) {
+       $InputObject | Write-DbaDataTable -SqlInstance $SqlInstanceSmo  -Database $Database -Table $Table -Schema $Schema -AutoCreateTable -Truncate:$Truncate
```
This way we will end with a HEAP with no data compression and all text/string values will become nvarchar(max)
![image](https://user-images.githubusercontent.com/19521315/80869189-dd0fba80-8c96-11ea-8225-62e5e8cae49c.png)

I would recommend having a CLUSTERED index on date column with DATA_COMPRESSION = PAGE.
If destination SQL Server instance isn't 2016 SP1 or higher or not enterprise edition, we need to fall back to default with no compression. 

see my comment on line 68

ClaudioESSilva commented on functions/Write-DbcTable.ps1:
```diff
@@ -0,0 +1,75 @@
+
+.DESCRIPTION
+After running Invoke-DbcCheck (With PassThru) and converting it to a datatable with Convert-DbcResult, this command
+will writet the results to a database table and will also write the current Checks to another table called dbachecksChecks
```
There is an extra 't' at the end of "write" word.

SQLDBAWithABeard replied:
Was already fixed (I have a code speller extension installed :-)

ClaudioESSilva commented on functions/Convert-DbcResult.ps1:
```diff
@@ -0,0 +1,118 @@
+        else {
+            $Database = $null
+        }
+        $Date = Get-Date # -Format "yyyy-MM-dd"
```
Should this be the datetime of the import or related with the run?
Should we have both?

SQLDBAWithABeard replied:
I dont think so, I only think that the time of the check is important

ClaudioESSilva replied:
I agree but the get-date will be the date of the object conversion and then becomes the import date.
Tbh i don't remember to see a date time of the run for each test as result of the pester test. Am I right?

SQLDBAWithABeard commented:
GRRRRRRRR

![image](https://user-images.githubusercontent.com/6729780/80992376-50e5ca80-8e31-11ea-8d63-b1c75e28e339.png)

So I was looking at handling the couldn't connects to be a little nicer and now I realise that the name of every single test where we cant connect is "Couldn't connect to" 

which is correct

but now I need to work out how to filter those out of the results in the PowerBi 

so I am walking away for a while

jpomfret commented on functions/Write-DbcTable.ps1:
```diff
@@ -0,0 +1,114 @@
+The datatable from Convert-DbcResult
+
+.PARAMETER Table
+The name of the table for the results - will be created if it doesnt exist by default it will be CheckResults
```
```suggestion
The name of the table for the results - will be created if it doesn't exist by default it will be CheckResults
```

SQLDBAWithABeard replied:
Thank you :-)

jpomfret commented on functions/Write-DbcTable.ps1:
```diff
@@ -0,0 +1,114 @@
+The name of the table for the results - will be created if it doesnt exist by default it will be CheckResults
+
+.PARAMETER Schema
+the schema for the table - defaults to dbo
```
```suggestion
The schema for the table - defaults to dbo
```

SQLDBAWithABeard replied:
thank you

jpomfret commented on functions/Write-DbcTable.ps1:
```diff
@@ -0,0 +1,114 @@
+    else {
+        if ($PSCmdlet.ShouldProcess("$schema.$database" , "Create a new table called $table ")) {
+            # If specified table does not exists, create with specific datatypes to avoid nvarchar(max)
+            $sqlTableCreation = @"
```
Really appreciate defining the table ahead of time like this... makes for happy Dbas!

SQLDBAWithABeard replied:
Yes, that was @ClaudioESSilva (although he missed that his own checks for FKs make the name too long :-D )
Be good to do the same for the Checks table or not so much as it has less than 200 rows?

SQLDBAWithABeard replied:
Thank you :-)

ClaudioESSilva replied:
Should we truncate the Name on the result object result to the 600 chars? We never know when someone is trying to telling a story using FK names 😂

SQLDBAWithABeard replied:
I dont know, I Adventureworks got it up to 320 so I nearly doubled it
Lets do this properly.

The title is 

Database $($psitem.Database) Check Constraint $($psitem.Name) on table $($psitem.Parent) should be trusted on $($psitem.SqlInstance)

Database can be 128
Check constraint can be 128
table can be 128
instance can be 16
Host can be 15 + \ = 16
Rest of title is 58

128 + 128 + 128 + 16 + 16 + 58 = 474

So 600 gives us room to add more detail there if needed - Be worth putting a Pester Test to ensure that the names are not too long when generated

jpomfret commented on functions/Start-DbcPowerBi.ps1:
```diff
@@ -1,25 +1,29 @@
 
-.PARAMETER Path
-The location of the pbix file. "$script:ModuleRoot\bin\dbachecks.pbix" by default.
+.PARAMETER FromDatabase
```
Not sure where to leave this comment, as I couldn't seem to on the pbix file.

The drillthrough seems to default to a date of 4/14/2020 so doesn't show any results until I check today's date. Wondering if that can be passed through? Way out of my depth regarding PowerBI Knowledge here :) 
![image](https://user-images.githubusercontent.com/981370/81420311-b5fa3280-9147-11ea-9d78-b1662d929501.png)

SQLDBAWithABeard replied:
Hmm, I need to think about htat, also not a PowerBi expert, but I am going to clear out the results and start with an empty file to see if that will make a difference

SQLDBAWithABeard replied:
Thank you :-)

SQLDBAWithABeard replied:
Yup, clearing the data from the template file has fixed that

![image](https://user-images.githubusercontent.com/6729780/81469704-60249980-91de-11ea-8483-96a7f808463b.png)
